test/memory.cpp

```diff
@@ -747,3 +749,290 @@ TEST(BufferInfo, UnlockCpp) {
+
+    // alloc
+    af_memory_manager_set_alloc_fn(manager, alloc_fn);
+    af_memory_manager_set_allocated_fn(manager, allocated_fn);
+    af_memory_manager_set_unlock_fn(manager, unlock_fn);
+    // utils
+    af_memory_manager_set_garbage_collect_fn(manager, garbage_collect_fn);
+    af_memory_manager_set_print_info_fn(manager, print_info_fn);
+    af_memory_manager_set_usage_info_fn(manager, usage_info_fn);
+    // user lock/unlock
+    af_memory_manager_set_user_lock_fn(manager, user_lock_fn);
+    af_memory_manager_set_user_unlock_fn(manager, user_unlock_fn);
+    af_memory_manager_set_is_user_locked_fn(manager, is_user_locked_fn);
+    // limits and step size
+    af_memory_manager_set_get_mem_step_size_fn(manager, get_mem_step_size_fn);
+    af_memory_manager_set_get_max_bytes_fn(manager, get_max_bytes_fn);
+    af_memory_manager_set_get_max_buffers_fn(manager, get_max_buffers_fn);
+    af_memory_manager_set_set_mem_step_size_fn(manager, set_mem_step_size_fn);
+    af_memory_manager_set_check_memory_limit_fn(manager, check_memory_limit_fn);
+    // ocl
+    af_memory_manager_set_add_memory_management_fn(manager,
+                                                   add_memory_management_fn);
+    af_memory_manager_set_remove_memory_management_fn(
+        manager, remove_memory_management_fn);
+
+    af_set_memory_manager(manager);
+    {
+        size_t aSize = 8;
+
+        void *a = af::alloc(8, af::dtype::f32);
+        ASSERT_EQ(payload->table.size(), 1);
+        ASSERT_EQ(payload->table[a], aSize * sizeof(float));
+
+        auto b = af::randu({2, 2});
+
+        // Usage info
+        size_t allocBytes, allocBuffers, lockBytes, lockBuffers;
+        af::deviceMemInfo(&allocBytes, &allocBuffers, &lockBytes, &lockBuffers);
+        ASSERT_EQ(allocBytes, aSize * sizeof(float) + b.bytes());
+        ASSERT_EQ(allocBuffers, 2);
+        ASSERT_EQ(lockBytes, aSize * sizeof(float) + b.bytes());
+        ASSERT_EQ(lockBuffers, 2);
+
+        af::free(a);
+
+        af::deviceMemInfo(&allocBytes, &allocBuffers, &lockBytes, &lockBuffers);
+        ASSERT_EQ(allocBytes, aSize * sizeof(float) + b.bytes());
+        ASSERT_EQ(allocBuffers, 2);
+        ASSERT_EQ(lockBytes, b.bytes());
+        ASSERT_EQ(lockBuffers, 1);
+
+        ASSERT_EQ(payload->table.size(), 2);
+    }
+
+    // gc
+    af::deviceGC();
+    ASSERT_EQ(payload->table.size(), 0);
+
+    // printInfo
+    std::string printInfoMsg = "testPrintInfo";
+    int printInfoDeviceId    = 0;
+    af::printMemInfo(printInfoMsg.c_str(), printInfoDeviceId);
+    ASSERT_EQ(printInfoMsg, payload->printInfoStringArg);
+    ASSERT_EQ(printInfoDeviceId, payload->printInfoDevice);
+
+    // step size
+    size_t stepSizeTest = 64;
+    af::setMemStepSize(stepSizeTest);
+    ASSERT_EQ(af::getMemStepSize(), stepSizeTest);
+    ASSERT_EQ(stepSizeTest, payload->memStepSize);
+
+    ASSERT_EQ(payload->table.size(), 0);
+    af_release_memory_manager(manager);
+}
```

Comment:
Finally: Something like this should eventually be added to the examples also, where we show how to write a custom memory manager.